### PR TITLE
require non-empty directive locations

### PR DESCRIPTION
**Pull request by jbellenger**
It's possible to programmatically create a `GraphQLDirective` without setting it's valid locations. These missing locations are not validated at schema creation time either, allowing the schema without directive locations to be rendered with`SchemaPrinter`, producing SDL like:
```
directive @foo on
directive @bar on
```

Trying to reparse this rendered sdl fails with an InvalidSyntaxError.

This PR addresses this by requiring that at the time that a `GraphQLDirective` is created, it must have a non-empty set of locations configured. This is similar to existing validations for name and arguments. As a result of this new requirement, I had to update a number of tests that create directives without locations.

There are 2 spec-related aspects to this:
1. Does the spec actually require a directive location?
It's not clear to me that it does. The [grammar](https://spec.graphql.org/October2021/#DirectiveLocations) part of DirectiveLocations (which I don't fully grok) seems ambiguous. I also did a shallow check of graphql-js and didn't find anything indicating a requirement for directive locations. 
If they are indeed *not* required, then this PR may be over-reaching by requiring non-empty locations in the GraphQLDirective constructor

2. The spec requires that [directives be applied in valid locations](https://spec.graphql.org/October2021/#sec-Directives-Are-In-Valid-Locations)
This suggests that some of the unit tests in this PR, which create and use directives that have no validLocations, are maybe a little lax.
Maybe this is fine at the level of individual schema objects, so long as they're eventually validated at schema creation? Feedback welcome!

### CODE REVIEW

**Comment by bbakerman** [comment]
> Does the spec actually require a directive location?

My reading of that spec says there are required at parse time

````
DirectiveDefinition : 
    Description(opt) directive @Name Arguments Definition(opt) repeatable(opt) on DirectiveLocations
```

DirectiveLocations is required as per spec and they should be be enforced.  Remember early graphql-java was a very direct port of graphql-js so its possible if they missed it, we missed it.

**Comment by bbakerman** [comment]
> This suggests that some of the unit tests in this PR, which create and use directives that have no validLocations, are maybe a little lax.

This is a perrinial challenge for us - that is we create objects in unit tests that would never present in real life production

```
def directive = GraphQLDirective.newDirective().name("cached").argument(arg).build()
```

Thats just lazy code and because we never had an assert, then this lazy code passed.  I think you change helps this.

File: src/test/groovy/graphql/TestUtil.groovy [hunk 1]
```diff
@@ -2,32 +2,11 @@ package graphql
...
-import graphql.schema.GraphQLType
-import graphql.schema.TypeResolver
-import graphql.schema.idl.RuntimeWiring
-import graphql.schema.idl.SchemaGenerator
-import graphql.schema.idl.SchemaParser
-import graphql.schema.idl.TestMockedWiringFactory
-import graphql.schema.idl.TypeRuntimeWiring
-import graphql.schema.idl.WiringFactory
+import graphql.schema.*
+import graphql.schema.idl.*
```
Comment (bbakerman) [inline src/test/groovy/graphql/TestUtil.groovy:9]: Take out the `.*` imports - we dont like them.  You may have to change your IDE settings when working on the graphql code base

File: src/test/groovy/graphql/TestUtil.groovy [hunk 2]
```diff
@@ -194,13 +173,19 @@ class TestUtil {
...
                 .definition(definition)
                 .build()
     }
 
-    static GraphQLDirective mockDirective(String name) {
-        newDirective().name(name).description(name).build()
+    static GraphQLDirective mockDirective(String name, DirectiveLocation location, GraphQLArgument arg = null) {
+        def b = newDirective().name(name).description(name).validLocation(location)
+        if (arg != null) b.argument(arg)
+        b.build()
```
Comment (bbakerman) [inline src/test/groovy/graphql/TestUtil.groovy:188]: Now I know you didn't name this method and it was called mock buty its not a mock.  Its a directive so it makes all that other code look weirder

Can we rename this as 

* directive()
* mkDirective()

**Review by bbakerman — CHANGES_REQUESTED** [review]
I approve subject to the asked for changes

**Comment by jbellenger** [comment]
> My reading of that spec says there are required at parse time

I think you're right, but I have low confidence. I think it's the recursive definition of DirectiveLocations that is confusing
```
DirectiveLocations : 
  DirectiveLocations | DirectiveLocation |opt DirectiveLocation
```
To put this into plain english, I think that it is supposed to be one or more DirectiveLocation elements joined by pipes.

I'll put a PR up on graphql-spec, I think the grammar could be clearer and would benefit from an explanation.

> I approve subject to the asked for changes

I'll make these changes, thanks for reviewing!

**Review by bbakerman — APPROVED** [review]

**Comment by bbakerman** [comment]
> Also, I wanted to call out that I think this PR might be considered a breaking change -- I'm not sure what that means for merging.

I guess is breaking in the sense that if you previously built invalid directives - they now wont build.  I feel this is ok

**Pull request merged**